### PR TITLE
Add tests for AnyPickler/AnyUnpickler

### PR DESCRIPTION
**Pull request by jvican**
- AnyUnpickler works if `HybridRuntime` is not used. But as it's in the
  context, it doesn't. AnyPickler doesn't work in any case.
- This could be used to overcome the limitations pointed out in #401.

### CODE REVIEW

**Comment by jvican** [comment]
These tests show how the current state of `AnyPickler` and `AnyUnpickler` doesn't work. As this looks like something easy to solve, do I have your permission to look into it @jsuereth?
/cc @phaller @heathermiller

**Comment by jvican** [comment]
BTW, just for the record, `AnyPickler` fails with a different error when `HybridRuntime` is not used: it actually doesn't serialize correctly `Box` and outputs a `()` and some more garbage.

**Comment by phaller** [comment]
@jvican How about creating a branch? F-P could then build on this branch until it's merged into 0.11.x.

**Comment by jvican** [comment]
That's actually my plan @phaller :smile: I'm first working on the fix for this since I've spotted the error. I'll open an issue and create a PR later! Afterwards, we'll use it in a branch of fp, although we'll also need to touch spores-pickling's PR.

**Comment by phaller** [comment]
@jvican sounds good! :+1:

**Comment by jsuereth** [comment]
@jvican please do!  I'd love to see what you come up with.  I'll be working on finishing up the FastTypeTag cleanup work, including starting to handle the "tricky" types that are commented out in the test.   Hopefully we don't have too many merge conflicts.  Ping me if you find issues with the FastTypeTag cleanups.

File: core/src/test/scala/pickling/run/anypickler-unpickler.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+class AnyPicklerUnpickler extends FunSuite {
+
+  import PicklingProtocol._
+
+  test("AnyPickler should work") {
+
+    val b = Box(List(1,2,3,4,5))
+    val containerPickler = implicitly[Pickler[List[Int]]]
+    val pickled = b.pickle.value
```
Comment (jsuereth) [inline core/src/test/scala/pickling/run/anypickler-unpickler.scala:29]: Note, you'll need to actually register the pickler with the hybrid runtime (currently) so this is expected to fail.

Specifically:

``` scala
scala.pickling.internal.currentRuntime.picklerRegistry.registerPickler(containerPickler)
```

Then, AnyPickler should be able to work with hybrid mode.

File: core/src/test/scala/pickling/run/anypickler-unpickler.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    import scala.pickling.pickler.AnyUnpickler
+    val format = implicitly[PickleFormat]
+    val reader = pickleFormat.createReader(JSONPickle(pickled))
+    val tag1 = reader.beginEntry()
+    val unpickled = AnyUnpickler.unpickle(tag1, reader)
```
Comment (jsuereth) [inline core/src/test/scala/pickling/run/anypickler-unpickler.scala:51]: Similarly, you'll need to register the unpickler.

As we discussed, I'd like for this registering to be automated when a pickler is generated.  If you'd like to discuss more, ping me in gitter.

**Comment by jsuereth** [comment]
@jvican Just a warning, I'm going to "squash" my commits in the FastTypeTag CL before committing, as I have a lot of loose/extraneous commits in there.

File: core/src/test/scala/pickling/run/anypickler-unpickler.scala (hunk 1 above)
Comment (jvican) [inline core/src/test/scala/pickling/run/anypickler-unpickler.scala:29]: Sorry for the delay of my reply. Yes, I realised that, but even registering the pickler I believe there's no possible way `AnyPickler` would work. The major problem is that `AnyPickler` deals with generic types that are erased as if they were `Any` and therefore the key that uses for looking up in the map is `scala.pickling.runtime.Box[scala.Any]` which does not match `scala.pickling.runtime.Box[scala.collection.immutable.List[scala.Int]]`.

Even if we register the tag with the `Any` keyword, the value mapped to that key would not work for other generic types of `Box`. The only way to solve this problem would be have access to the actual types behind `Any`, but again, the class name will only return the erased type of Box.

I've also tried other workarounds that give no success.

**Comment by jvican** [comment]
Go ahead @jsuereth :+1: Thanks for the heads up.

**Pull request closed without merging**